### PR TITLE
first draft for BN support in nn.Sequential

### PR DESCRIPTION
**Pull request by paganpasta**
Added the StatefulLayer as per discussion https://github.com/patrick-kidger/equinox/issues/448

One doubt I have is the `inference` parameter, it is not utilised by other norms. 
If BN needs to be easily interchangable with other norms, we might want to either remove `inference` parameter from BN and only rely on the self.inference flag or
add `inference` params to other normalisation layers as well. 

Also, do all Stateful layers utilise the `inference` flag?

### CODE REVIEW

File: equinox/nn/_batch_norm.py [hunk 1]
```diff
@@ -10,7 +11,29 @@
 from ._stateful import State, StateIndex
 
 
-class BatchNorm(Module):
+class _StatefulLayer(Module):
```
Comment (patrick-kidger) [inline equinox/nn/_batch_norm.py:14]: Let's put this in `nn/_stateful.py`, and additionally make it public API. (i.e. provide it in `__init__.py`). This should be something an end user could override in principle.

(I'm happy to hold off on documenting it for now, though.)

File: equinox/nn/_batch_norm.py [hunk 2]
```diff
@@ -10,7 +11,29 @@
...
+    current state as arguments and returns the output data and updated state.
+    """
+
+    @abc.abstractmethod
+    def __call__(
+        self,
+        x,
+        state: State,
+        *,
+        key: Optional["jax.random.PRNGKey"] = None,  # pyright: ignore
```
Comment (patrick-kidger) [inline equinox/nn/_batch_norm.py:30]: This should now be `Optional[jaxtyping.PRNGKeyArray]`, see the other examples elsewhere in Equinox.

File: equinox/nn/_composed.py [hunk 3]
```diff
@@ -136,32 +140,62 @@ class Sequential(Module):
         Activation functions can be added by wrapping them in [`equinox.nn.Lambda`][].
     """
 
-    layers: tuple[_SequentialLayer, ...]
+    layers: tuple[Union[_SequentialLayer, _StatefulLayer], ...]
```
Comment (patrick-kidger) [inline equinox/nn/_composed.py:143]: I think I'd suggest just removing the `_SequentialLayer` protocol. IMO with this update it becomes more confusing than it is useful.

Comment (paganpasta) [reply equinox/nn/_composed.py:143]: Wouldn't that imply that layers now only accepts `StatefulLayer` ?

Comment (patrick-kidger) [reply equinox/nn/_composed.py:143]: I mean to remove both from type annotations: just have `layers: tuple`.

File: equinox/nn/__init__.py [hunk 4]
```diff
@@ -34,4 +34,8 @@
 )
 from ._rnn import GRUCell as GRUCell, LSTMCell as LSTMCell
 from ._spectral_norm import SpectralNorm as SpectralNorm
-from ._stateful import State as State, StateIndex as StateIndex
+from ._stateful import (
+    State as State,
+    StatefulLayer,  # noqa: F401
```
Comment (paganpasta) [inline equinox/nn/__init__.py:39]: A bit unsure about this.

Comment (patrick-kidger) [reply equinox/nn/__init__.py:39]: You need `StatefulLayer as StatefulLayer`. The `X as X` pattern is used to include that it's public API, despite being imported into this file.

**Comment by patrick-kidger** [comment]
Also `SpectralNorm` is also stateful (the only other stateful layer built-in to Equinox), so it should be marked as such.

File: equinox/nn/_composed.py [hunk 5]
```diff
@@ -136,32 +139,58 @@ class Sequential(Module):
...
+                x, state = layer(x, state=state, key=key)
+            else:
+                x = layer(x, key=key)
+        if state is sentinel:
+            return x
+        else:
+            return x, state
 
-    def __getitem__(self, i: Union[int, slice]) -> _SequentialLayer:
+    def __getitem__(self, i: Union[int, slice]):
```
Comment (paganpasta) [inline equinox/nn/_composed.py:193]: @patrick-kidger Not sure about the return type here.

Comment (patrick-kidger) [reply equinox/nn/_composed.py:193]: I think it's always going to be a `Module` -- it's indexing or slicing `self.layer: Sequence[Module]`.

**Comment by paganpasta** [comment]
Surprisingly, the pre-commit passed on my local.

**Comment by patrick-kidger** [comment]
Oh sorry, I only just saw your comment here. I think just adding a `# pyright: ignore` is fine here.
(Can you squash your commits together into just one commit please?)

File: equinox/nn/_composed.py [hunk 6]
```diff
@@ -136,32 +139,58 @@ class Sequential(Module):
         Activation functions can be added by wrapping them in [`equinox.nn.Lambda`][].
     """
 
-    layers: tuple[_SequentialLayer, ...]
+    layers: tuple
 
-    def __init__(self, layers: Sequence[_SequentialLayer]):
+    def __init__(self, layers: Sequence[Module]):
```
Comment (patrick-kidger) [inline equinox/nn/_composed.py:144]: `Module` isn't quite accurate here -- we can put non-modules here too. I think `Callable` should be used here. (`from collections.abc import Callable`)

Comment (patrick-kidger) [reply equinox/nn/_composed.py:144]: Also, now that `_SequentialLayer` is unused, can it be deleted?

File: equinox/nn/_composed.py [hunk 7]
```diff
@@ -136,32 +139,58 @@ class Sequential(Module):
...
+        key: Optional[PRNGKeyArray] = None,
+    ) -> Union[Array, tuple[Array, State]]:
         """**Arguments:**
 
-        - `x`: Argument passed to the first member of the sequence.
-        - `key`: A `jax.random.PRNGKey`, which will be split and passed to every layer
-            to provide any desired randomness. (Optional. Keyword only argument.)
+        - `x`: passed to the first member of the sequence.
+        - `state`: If provided, then it is passed on to any layer which implements
+            `equinox._batch_norm._StatefulLayer`.
```
Comment (patrick-kidger) [inline equinox/nn/_composed.py:168]: "then it is passed to any layer which subclasses `eqx.nn.StatefulLayer`, which is used purely to indicate that this layer accepts and returns an extra `state` argument"

File: equinox/nn/_composed.py [hunk 8]
```diff
@@ -136,32 +139,58 @@ class Sequential(Module):
...
+                x, state = layer(x, state=state, key=key)
+            else:
+                x = layer(x, key=key)
+        if state is sentinel:
+            return x
+        else:
+            return x, state
 
-    def __getitem__(self, i: Union[int, slice]) -> _SequentialLayer:
+    def __getitem__(self, i: Union[int, slice]) -> Module:
```
Comment (patrick-kidger) [inline equinox/nn/_composed.py:193]: Likewise the return annotation here should actually be `Callable`, not `Module`, I've realised.

File: equinox/nn/_stateful.py [hunk 9]
```diff
@@ -145,3 +146,24 @@ def tree_unflatten(cls, keys, values):
...
+
+class StatefulLayer(Module):
+    """An abstract base class representing a stateful layer within a neural network.
+
+    Stateful layers maintain internal state across calls, which can be useful for tasks
+    such as BatchNormalisation.
+
+    Subclasses must implement the `__call__` method that takes input data and the
+    current state as arguments and returns the output data and updated state.
+    """
```
Comment (patrick-kidger) [inline equinox/nn/_stateful.py:159]: Can you (a) mention that right now this is used solely to declare the API for use with Sequential, and (b) add this base class to the documentation?

**Comment by patrick-kidger** [comment]
And merged! Thank you for the implementation, I'm glad to have it in.
